File: szakdoliV2/Drawable/Base/Command.py
# -*- coding: utf-8 -*-

from abc import abstractmethod
import warnings

from Drawable.Base import AbstractDrawable

from Drawable.Rectangle import Rect
from Drawable.Sprite import Spirte
from Drawable.TextIcon import TextIcon

class Command(AbstractDrawable):

    def __init__(self, x=None, y=None, w=None, h=None, vec2_pos=None, size=None):
        super(Command, self).__init__(x, y, w, h, vec2_pos, size)
        self.mainRect = Rect(self.x, self.y, self.w, self.h, width=1, transparent=False)
        self.sprite = None
        self.imagePath = None
        self.texticon = None
        self.keycode = None
        self.keycode_pad = 0

    def execute_command(self):
        raise RuntimeError("Event didn't binded")

    def IsInside(self, position):
        return self.x <= position[0] and self.x + self.w >= position[0] and self.y <= position[1] and self.y + self.h >= position[1]

    def DrawObject(self, screen):
        if not (self.sprite or self.texticon):
            self.LoadSprite()
        self.mainRect.DrawObject(screen)
        if self.sprite:
            self.sprite.DrawObject(screen)
        if self.texticon:
            self.texticon.DrawObject(screen)

    def LoadSprite(self):
        if self.imagePath is None and self.keycode is not None:
            self.texticon = TextIcon(self.x, self.y, self.w, self.h, keycode=self.keycode)
            self.texticon.SetPaddning(self.keycode_pad)
        elif self.imagePath is not None and self.keycode is None:
            self.sprite = Spirte(self.x, self.y, self.w, self.h, imgpath=self.imagePath)
        else:
             self.sprite = Spirte(self.x, self.y, self.w, self.h)
             warnings.warn("Too many argument")

    def SetPosition(self, x, y):
        self.x = x
        self.y = y
        self.mainRect.SetPosition(x,y)
        if self.sprite:
            self.sprite.SetPosition(x,y)
        if self.texticon:
            self.texticon.SetPosition(x,y)

    def deltapos(self, position):
        return position[0] - self.x, position[1] - self.y

    def drag(self, mouseposition):
        x = mouseposition[0] - self.delta[0]
        y = mouseposition[1] - self.delta[1]
        self.SetPosition(x, y)

    def setDelta(self, mouseposition):
        self.delta = self.deltapos(mouseposition)

    def UnloadIcon(self):
        self.texticon = None

    def Bind(self, function_pointer):
        self.execute_command = function_pointer

    def SetAccentColor(self, accent):
        self.mainRect.SetAccentColor(accent)

    def SetKeyCodePadding(self, x):
        self.keycode_pad = x

File: docs/Drawable/Base/Command.js
import { AbstractDrawable } from "./AbstractDrawable.js";
import { Rect } from "../Rectangle.js";
import { TextIcon } from "../TextIcon.js";
import { Sprite } from "../Sprite.js";
import { IMG } from "../../System/Constants.js";

export class Command extends AbstractDrawable {
    constructor(x = null, y = null, w = null, h = null) {
        super(x, y, w, h);
        this.mainRect = new Rect(x, y, w, h, undefined, 1, undefined, false);
        this.sprite = null;
        this.imagePath = null;
        this.textIcon = null;
        this.keyCode = null;
        this.keyCodePad = 0;
    }

    ExecuteCommand() {
        console.log("Event didn't bound");
    }

    IsInside(position) {
        return this.x <= position[0] && this.x + this.w >= position[0] && this.y <= position[1] && this.y + self.h >= position[1];
    }

    DrawObject(screen) {
        if (! (this.sprite || this.textIcon)){
            this.LoadSprite()
        }
        this.mainRect.DrawObject(screen);
        if (this.sprite) {
            this.sprite.DrawObject(screen);
        }
        if (this.textIcon) {
            this.textIcon.DrawObject(screen);
        }
    }

    LoadSprite() {
        if (this.imagePath == null && this.keyCode != null) {
            this.textIcon = new TextIcon(this.x, this.y, this.w, this.h, this.keyCode, undefined);
            this.textIcon.SetPadding(this.keyCodePad)
            
        } else if (this.imagePath != null && this.keyCode == null) {
            this.sprite = new Sprite(this.x, this.y, this.w, this.h, this.imagePath);
        } else {
            this.sprite = new Sprite(this.x, this.y, this.w, this.h);
            console.log("Too many argument");
        }
    }
    
    SetPosition(x, y){
        this.x = x;
        this.y = y;
        this.mainRect.SetPosition(x,y);
        if(this.sprite){
            this.sprite.SetPosition(x,y);
        }

        if(this.textIcon){
            this.textIcon.SetPosition(x,y);
        }
    }

    Deltapos(position){
        return [position[0] - this.x, position[1] - this.y];
    }

    Drag(mouseposition) {
        var x = mouseposition[0] - this.delta[0];
        var y = mouseposition[1] - this.delta[1];
        this.SetPosition(x, y);
    }

    SetDelta(mouseposition){
        this.delta = this.Deltapos(mouseposition);
    }

    UnloadIcon(){
        this.textIcon = null;
    }

    Bind(function_pointer){
        this.ExecuteCommand = function_pointer;
    }

    SetAccentColor(accent){
        this.mainRect.SetAccentColor(accent);
    }

    SetKeyCodePadding(x){
        this.keyCodePad = x;
    }

}